batches: add batch changes badge and filtered page for repo

courier-new opened the pull request:
Closes #20309.

Hello, fellow batchers, and welcome to episode 1 of Repo Batch Changes Badge, otherwise known as RBCB! The synopsis is simple but compelling:

> An unsuspecting Sourcegraph user discovers a treasure trove hidden right beside their repositories. 
What will they do when they uncover... batch changes??

In this pilot episode, we'll introduce you to the full cast of characters and tease the long-running plot, essentially a whole new mode of interacting with existing batch changes from a repository. And we promise there's no cliffhanger ending.

## The cast
(aka the main things I'd love for you to pay attention to in this early review 😛)

<table><tr><td><h3>The Batch Changes Badge</h3></td><td><img src="https://user-images.githubusercontent.com/8942601/123499652-1323a280-d5ed-11eb-880d-72ba586044bf.png" width="200"></td></tr></table>

Your charming, enigmatic protagonist who only shows up when there's something exciting happening.

Though I haven't finished hooking this up to GraphQL yet, I was imagining that for visibility we show this badge whenever either

- The number of open changesets on this repository is > 0, or
- The number of merged changesets on this repository is > 0,

as both would provide something for the user to look at when they get to the new repo page and draw positive attention to batch changes, even if it's not much.

<table><tr><td><h3>The <a href="https://5f0f381c0e50750022dc6bf7-tylinnzafb.chromatic.com/?path=/story/web-batches-batchchangerepopage--list-of-batch-changes" target="_blank">Repo Batch Changes Page</a></h3></td><td><img src="https://user-images.githubusercontent.com/8942601/123499805-19664e80-d5ee-11eb-8c96-32482a39b6c9.png" width="300"></td></tr></table>

They've got lots to show and tell, but you feel like you've seen it all before...

(...that's because you have -- I basically created a mash-up of components from the `BatchChangeListPage` and the `BatchChangeDetailsPage`)

The screenwriters couldn't totally agree on the character design and landed somewhere between [this](https://www.figma.com/file/Wz9saTgmxwn1vgD3zPjTh8/expansion-ideas?node-id=35%3A812) and [this](https://www.figma.com/file/Wz9saTgmxwn1vgD3zPjTh8/expansion-ideas?node-id=42%3A2690); they mostly just wanted the character to feel natural amidst the characters it draws inspiration from.

The search + filters, bulk actions, and stats aren't fully wired up to GraphQL yet either because I wanted to get feedback on if they all make sense here or not first. The interesting thing to note about the search and filters is that they'd be applied to the individual changesets, not the parent batch changes, as I felt it made more sense that a user would want to operate on the specific changesets that are attached to this repository from here, and not so much the batch changes they originated from. This will also make wiring these up slightly more complicated, but my plan is to apply them to the nested changesets query and then hide the batch changes from the list that don't have any changesets matching the filters.

I will probably initially release this without bulk action support because the bulk action mutations currently require the `batchChange` id as an argument in order to check permissions, which gets more complicated if our bulk action is happening across multiple batch changes. I can just hide the checkboxes initially.

I have chosen for now to omit the Open/Closed label on the batch change itself, as well, as it doesn't feel like it adds much to this page. I'm also not yet sold on the visual indicator for the batch change that we had sketched up here on the right:

![image](https://user-images.githubusercontent.com/8942601/123500062-18ceb780-d5f0-11eb-977c-e81328ef3330.png)

If anyone feels particularly strongly about either of those elements I'm happy to experiment though.

<table><tr><td><h3>The <a href="https://github.com/sourcegraph/sourcegraph/pull/22372/files?file-filters%5B%5D=.graphql#diff-e81543ff15f2f5ee599339869bf4adacf439d298eb13eb80909c1aba6404030f">new Repo query fields</a></h3></td><td><img src="https://user-images.githubusercontent.com/8942601/123500193-1caf0980-d5f1-11eb-8d3d-2d552017778d.png" width="300"></td></tr></table>

They're the new kid in town who is interested in one thing and one thing only: their repository!

I basically replicated the `batchChanges`, `changesetsStats`, and `diffStat` fields from elsewhere on the schema and tailored them to the details these pages want for a specific repository. The `changesetStats` will power both the Batch Changes Badge as well as the stats row at the top of the Repo Batch Changes Page.

With the exception of the bulk action support described above, there were no additional mutation requirements needed for this new badge and page.

## On the next episode...

(aka questions that I would love your input on!!)

1. ✅ How do repo permissions work? Is it possible for a user to view this (or any other) repo page on Sourcegraph for a repo they don't have access to on the codehost? Would they just see a bunch of `HiddenExternalChangeset`s if they did? **A: It shouldn't be possible.**
2. ✅ Does having search + filters for changesets (vs. for the batch changes themselves) make sense on this list? **A: Exclude search + filters from initial release.**
3. ✅ Which changesets stats should we include on the row at the top of the page? Currently I have total, open, unpublished, closed, and merged. **A: Seems like a good set.**
	3a. ✅ If a given stat has 0 of its type (e.g. 0 closed changesets), should I omit the stat from the row, or keep it and just show 0? **A: Keep it and show 0.**
4. ✅ Erik already gave me some pointers on handling the enterprise-in-OSS parts of this, is there anything else I missed here? **A: Looks good.**
5. I am generally erring on the side of duplicating code here over generalizing existing code or forming new abstractions when it would only serve two places. I'm going to do a lap of cleanup and assess if any of these places make sense to DRY up, but if anything jumps out at you as being better served that way, please suggest it!
6. ✅ What stories should we include for the Repo Batch Changes page? I currently have with a full list, with a list with hidden external changesets, and with no batch changes. **A: Can safely exclude the middle one, shouldn't be possible. No others needed.**


## Plot holes

(aka tracking the things that I'm still working on/cleaning up for this PR)

- [x] Wire up changeset stats queries on the frontend
- [x] Wire up diff stat query on the frontend
- [x] ~Wire up changeset search + filters query parameters~
- [x] ~Hide a batch change from the list if all of its changesets are hidden by the current search parameters~
- [x] Clean up repo permissions
- [x] Exclude archived changesets from queries, stats
- [ ] Write backend tests for new GraphQL queries
- [x] If page is empty, point user to batch changes getting started materials
- [x] Fix initial page title
- [ ] File issue to track enhancements: bulk action support, search + filters
- [ ] CHANGELOG entry

eseliger commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -2416,6 +2416,40 @@ type Repository implements Node & GenericSearchResultInterface {
     The star count the repository has in the code host.
     """
     stars: Int!
+
+    """
+    Stats on all the changesets that have been applied to this repository by batch
+    changes.
+    """
+    changesetsStats: RepoChangesetsStats!
+
+    """
+    A list of batch changes that have applied a changeset to this repository.
+    """
+    batchChanges(
+        """
+        Returns the first n batch changes from the list.
+        """
+        first: Int = 50
+        """
+        Opaque pagination cursor.
+        """
+        after: String
+        """
+        Only return batch changes in this state.
+        """
+        state: BatchChangeState
+        """
+        Only include batch changes that the viewer can administer.
+        """
+        viewerCanAdminister: Boolean
+    ): BatchChangeConnection!
+
+    """
+    A diff stat for all the changesets that have been applied to this repository
+    by batch changes.
+    """
+    batchChangesDiffStat: DiffStat!
```
This schema extension should also go into `batches.graphql`, otherwise the fields would appear in the OSS version and since the `EnterpriseResolvers.batchChangesResolver` is nil there it would panic. 

You can do something like this:

```graphql
extend type Repository {
    """
    Stats on all the changesets that have been applied to this repository by batch
    changes.
    """
    changesetsStats: RepoChangesetsStats!
    """
    A list of batch changes that have applied a changeset to this repository.
    """
    batchChanges(
        """
        Returns the first n batch changes from the list.
        """
        first: Int = 50
        """
        Opaque pagination cursor.
        """
        after: String
        """
        Only return batch changes in this state.
        """
        state: BatchChangeState
        """
        Only include batch changes that the viewer can administer.
        """
        viewerCanAdminister: Boolean
    ): BatchChangeConnection!
    """
    A diff stat for all the changesets that have been applied to this repository
    by batch changes.
    """
    batchChangesDiffStat: DiffStat!
}
```

courier-new replied:
Thank you for answering what would have been one of my other questions!! 😂 You mention it should **also** go in `batches.graphql`, though -- so I should duplicate it in both places? I guess I was expecting it would go either here or there, but not both.

eseliger commented on client/web/src/repo/tree/TreePage.tsx:
```diff
@@ -373,11 +374,33 @@ export const TreePage: React.FunctionComponent<Props> = ({
+                                            className="mb-3 test-tree-page-title"
+                                        />
+                                        {repo.description && <p>{repo.description}</p>}
+                                        <Link
+                                            className="btn btn-sm btn-outline-secondary d-flex align-items-center mb-1 text-uppercase"
+                                            to={`/${encodeURIPathComponent(repo.name)}/-/batch-changes`}
+                                            aria-label="Batch Changes"
+                                        >
+                                            <BatchChangesIcon className="mr-2" />
+                                            Batch Changes
+                                            <span
+                                                className="badge badge-success batch-change-badge d-flex flex-column ml-2"
+                                                data-tooltip="10 open changesets"
+                                            >
+                                                10
+                                            </span>
+                                            <span
+                                                className="badge badge-merged batch-change-badge ml-2"
+                                                data-tooltip="12 merged changesets"
+                                            >
+                                                12
+                                            </span>
+                                        </Link>
```
This should only be shown if `props.showBatchChanges` is true.

courier-new replied:
Ahhhh I was looking for something like this, thanks. Let the prop-drilling commence! 😅

eseliger commented on client/web/src/repo/routes.tsx:
```diff
@@ -63,6 +67,14 @@ export const repoContainerRoutes: readonly RepoContainerRoute[] = [
             </div>
         ),
     },
+    {
+        path: '/-/batch-changes',
+        render: context => (
+            <RepositoryGitDataContainer {...context} repoName={context.repo.name}>
+                <RepositoryBatchChangesArea {...context} />
+            </RepositoryGitDataContainer>
+        ),
+    },
```
This route should only be registered in enterprise - we cannot import from `../enterprise` in `./client/web/!enterprise`

eseliger replied:
Not sure if we have this pattern for repo already, but user and org settings pages have a concept for this, which you could ~steal~ get inspired by.

courier-new replied:
Concept ~stolen~ was very inspiring indeed! It looks like there was already a bit of scaffolding in place for this sort of thing with `repo`, even, that I missed before, so I was able to take advantage of that!

eseliger reviewed (commented):
Sorry for peeking 😬 I was curious how you solved the "enterprise in OSS" part and left a few comments :)

courier-new commented:
Hahaha I think the takeaway is, I *did not* in fact solve the "enterprise in OSS" problem. 😂 Thanks for the early feedback, Erik!! I was definitely going to ask about all these things anyway so you've saved me the trouble of all that. 😄

eseliger commented:
Haha sorry - I didn't mean to point that out! I was just pretty sure that this'll be the rough part of this topic, so I figured I'd take a look early :)

LawnGnome reviewed (commented):
This looks good! I reviewed the GraphQL and Go code a bit more in depth than the frontend for now, but I at least glanced at everything. Feels like we're well on the way to an excellent season finale.

Also, your PR description is :fire:.

I'll take on your questions, but others will probably also have opinions:

> How do repo permissions work? Is it possible for a user to view this (or any other) repo page on Sourcegraph for a repo they don't have access to on the codehost? Would they just see a bunch of HiddenExternalChangesets if they did?

It _shouldn't_ be possible for a user to view the page if they don't have access, but we still need to implement security checks in the resolvers to handle the case where a user doesn't have access to the repo in case they're doing naughty things directly via GraphQL. I've put a couple of comments in indicating where we might need to add checks.

> Does having search + filters for changesets (vs. for the batch changes themselves) make sense on this list?

Not to me, since the top level object here is the batch change. Given that we don't have any sort of search right now on the top level batch changes page, I'd probably omit search and filtering for now here.

> Which changesets stats should we include on the row at the top of the page? Currently I have total, open, unpublished, closed, and merged.

That feels reasonable to me. Do we want to split draft out? (I'd probably say no, but we may as well discuss it.)

> If a given stat has 0 of its type (e.g. 0 closed changesets), should I omit the stat from the row, or keep it and just show 0?

I would keep and show 0.

> Erik already gave me some pointers on handling the enterprise-in-OSS parts of this, is there anything else I missed here?

LGTM.

> I am generally erring on the side of duplicating code here over generalizing existing code or forming new abstractions when it would only serve two places. I'm going to do a lap of cleanup and assess if any of these places make sense to DRY up, but if anything jumps out at you as being better served that way, please suggest it!

It feels like this is pretty close to what we'd want right now: I note that you embedded the stats struct and split it, for example, which feels appropriate.

> What stories should we include for the Repo Batch Changes page? I currently have with a full list, with a list with hidden external changesets, and with no batch changes.

If we don't have search/filter/bulk operations, that probably covers it. This page feels like it's a gateway and a search tool more than a functional page in terms of manipulating batch changes, so its behaviour should be relatively simple to start with.

To pick up on one more thing:

> I will probably initially release this without bulk action support because the bulk action mutations currently require the batchChange id as an argument in order to check permissions, which gets more complicated if our bulk action is happening across multiple batch changes. I can just hide the checkboxes initially.

I definitely don't consider bulk operations to be in scope for this project, for what it's worth. Is there a far distant day in the future where I could see us adding them? Sure. Can I think of a likely user story that would require them? Honestly, not really.

eseliger commented on client/web/src/enterprise/batches/repo/BatchChangeRepoPage.story.tsx:
```diff
@@ -0,0 +1,83 @@
+    </EnterpriseWebStory>
+))
+
+// TODO: Is this possible?? Can a user view a repository if they don't have access to it on the code host?
```
No I don't think they can :)

mrnugget commented on enterprise/internal/batches/resolvers/resolver.go:
```diff
@@ -676,12 +676,46 @@ func (r *Resolver) BatchChanges(ctx context.Context, args *graphqlbackend.ListBa
+		return nil, err
+	}
+
+	stats, err := r.store.GetRepoChangesetsStats(ctx, repoID)
```
You can also add the authz conditions to the SQL query itself: https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+AuthzQueryConds%20file%3Abatches

That's newer than the approach Adam described (try to load a repo from the database, if you get notfound then you probably can't access it) which is why it's not used everywhere.

mrnugget commented on enterprise/internal/batches/store/batch_changes.go:
```diff
@@ -320,6 +321,34 @@ func getBatchChangeDiffStatQuery(opts GetBatchChangeDiffStatOpts, authzConds *sq
+	COALESCE(SUM(diff_stat_changed), 0) AS changed,
+	COALESCE(SUM(diff_stat_deleted), 0) AS deleted
+FROM changesets
+WHERE changesets.repo_id = %s
```
This also needs a `AND repo.deleted_at IS NULL` (see [the other places in which we add this condition](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+repo.deleted_at%20file%3Abatches)) because repositories can be soft-deleted (still in the database, but `deleted_at` set to a timestamp) and the diff stats would be off then.

courier-new replied:
Oh thanks for mentioning this, I was actually wondering about this but forgot to ask. I get that the diff stats would be off if I were aggregating this across multiple repositories, but since this diff stats query is specific to a single repo, might it actually make sense to omit the `deleted_at` check? I doubt there's a real use case for this, but if we have a repo that's been deleted and someone did _hypothetically_ request the diff stats for that specific repo, by its id, fully knowing it's been deleted, wouldn't it be strange if all the diff stats were just 0? Again I'm really not expecting someone to try this query for a deleted repo, but just curious what you think.

mrnugget replied:
> wouldn't it be strange if all the diff stats were just 0?

No, I don't think so. I think "nothing is returned" (meaning something like a 404) is the best case and "null data" the 2nd best case.

courier-new replied:
Okay, that makes sense! Thanks!

mrnugget commented on enterprise/internal/batches/store/batch_changes.go:
```diff
@@ -402,6 +433,14 @@ func listBatchChangesQuery(opts *ListBatchChangesOpts) *sqlf.Query {
+		preds = append(preds, sqlf.Sprintf(`EXISTS(
+			SELECT * FROM changesets
+			WHERE changesets.batch_change_ids ? batch_changes.id::TEXT
+			AND changesets.repo_id = %s
```
Same as above: we probably need authz checks here and a check for `repo.deleted_at`

mrnugget commented on enterprise/internal/batches/store/changesets.go:
```diff
@@ -282,6 +283,9 @@ func countChangesetsQuery(opts *CountChangesetsOpts, authzConds *sqlf.Query) *sq
 		preds = append(preds, authzConds)
 	}
+	if opts.RepoID != 0 {
+		preds = append(preds, sqlf.Sprintf("changesets.repo_id = %s", opts.RepoID))
```
I think we already join `repo` in this query, so we could to `repo.id = %s` here. I'm not sure whether that has any effect at all. Just wanted to point out that `repo` is available :)

courier-new replied:
Oh true true! I don't _think_ that would have any effect, do you think using `repo.id = %s` is clearer?

mrnugget replied:
Hehe, no, not necessarily. I was debating with myself whether to even post this comment or not, because I'm not sure if one is an improvement over the other. I think I personally would use `repo.id = %s` just to make clear that `repo` _is_ already joined for the reader. But that's _highly subjective_ and more emotional than anything else.

mrnugget reviewed (commented):
Looks great so far! Left just a few comments to hopefully help with future work.

eseliger commented on cmd/frontend/graphqlbackend/repository.go:
```diff
@@ -141,6 +141,22 @@ func (r *RepositoryResolver) CloneInProgress(ctx context.Context) (bool, error)
+func (r *RepositoryResolver) BatchChanges(ctx context.Context, args *ListBatchChangesArgs) (BatchChangesConnectionResolver, error) {
+	id := r.ID()
+	args.Repo = &id
+	return EnterpriseResolvers.batchChangesResolver.BatchChanges(ctx, args)
```
Yes and I think the check would be rather redundant. If it's nil, it should panic, and if it's nil and we don't do a check, it panics too. :D

courier-new replied:
That's a great observation. Your guess is probably better than mine, but that was also my operating assumption. 😅

courier-new commented on client/web/src/repo/RepoContainer.tsx:
```diff
@@ -92,6 +92,8 @@ export interface RepoContainerContext
 
     globbing: boolean
+
+    showBatchChanges: boolean
```
It took me little while to trace this, too, but yeah, it comes from the _server_ router, and is injected as prop into the root `SourcegraphWebApp` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@kr/batch-changes-badge/-/blob/client/web/src/enterprise/main.tsx#L59:13). That value it pulls from (`window.context.batchChangesEnabled`) comes from the server checking the current user's permissions [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@kr/batch-changes-badge/-/blob/cmd/frontend/internal/app/jscontext/jscontext.go#L143:1) on this `JSContext` struct, which gets exposed and attached to `window.context` on the web app along with a bunch of other environment/settings metadata somehow through the server router (some magic happening in `frontend/internal/app/ui/handlers`):

![image](https://user-images.githubusercontent.com/8942601/124176116-56559980-da63-11eb-90f6-0a4b9a6bb43f.png)

courier-new commented on enterprise/internal/batches/store/changesets.go:
```diff
@@ -927,6 +935,33 @@ WHERE
 `
 
+// GetRepoChangesetsStats returns statistics on all the changesets associated to the given repo.
+func (s *Store) GetRepoChangesetsStats(ctx context.Context, repoID api.RepoID) (*btypes.RepoChangesetsStats, error) {
+	authzConds, err := database.AuthzQueryConds(ctx, s.Handle().DB())
+	if err != nil {
+		return nil, errors.Wrap(err, "GetRepoChangesetsStats generating authz query conds")
+	}
+	q := getRepoChangesetsStatsQuery(int64(repoID), authzConds)
```
@mrnugget I also added the repo permission checks to this method for changeset stats on a repo, but let me know if you don't feel like it belongs here!

mrnugget replied:
No, belongs there :)

eseliger commented:
Actually, I don't think that having bulk actions on this page is a requirement.

courier-new commented:
> Actually, I don't think that having bulk actions on this page is a requirement.

Okay great, you and @LawnGnome seem to agree on that then, this was Adam's comment:

> I definitely don't consider bulk operations to be in scope for this project, for what it's worth. Is there a far distant day in the future where I could see us adding them? Sure. Can I think of a likely user story that would require them? Honestly, not really.

I intended to include it because we mention it in the ticket proposal:

> from the batch change repository page, the user can take actions on the changeset, either by navigating to the codehost, or by selecting one or many changesets and **trigger a bulk action.**

And we added the checkboxes + "Actions" menu to all the designs we initially brainstormed, but I think I'll omit it for now and we can file an issue to revisit if we get the sense that it'd actually be useful. 🙂

courier-new commented:
In light of our discussion about whether or not we could have broken this down better or turn several of the biggest items on our checklists into individual tasks, and since I've addressed all of the feedback on this draft PR, I'm actually going to close this in favor of opening several smaller piecemeal PRs instead, to pretend like I broke this up better from the start. 😉

malomarrec commented:
🥳 👏 👏 👏
